### PR TITLE
Struct & SIMD improvements

### PR DESCRIPTION
**Pull request by CarolEidt**
- Enable CSE of struct values when handle is available (and add code to get the handle of HW SIMD types)
- Don't require block nodes for SIMD assignments
- Don't set `GTF_GLOB_REF` on `GT_OBJ` if it is local
- Set `lvRegStruct` on promoted SIMD fields
- Add tests for #19910 (fixed with this PR) and #3539 & #19910 (fixed with #21314)
- Additional cleanup

### CODE REVIEW

File: src/jit/rationalize.h [hunk 1]
```diff
@@ -39,22 +39,22 @@ class Rationalizer : public Phase
     void RewriteSIMDOperand(LIR::Use& use, bool keepBlk);
 
     // Intrinsic related transformations
-    void RewriteNodeAsCall(GenTree**             use,
-                           ArrayStack<GenTree*>& parents,
-                           CORINFO_METHOD_HANDLE callHnd,
+    void RewriteNodeAsCall(GenTree**               use,
+                           Compiler::GenTreeStack& parents,
```
Comment (mikedn) [inline src/jit/rationalize.h]: Why the change?

Comment (mikedn) [reply src/jit/rationalize.h]: I'd keep `ArrayStack<GenTree*>`. It's not some convoluted template instantiation that warrants the use of a typedef and the `GenTreeStack` typedef is used only in 3 places so it's best not to add more (and remove it at some point).

**Comment by mrange** [comment]
👍 I really hope for this in dotnet 3.0 release.

**Comment by CarolEidt** [comment]
> I really hope for this in dotnet 3.0 release.

Working on it, and hope to get it in.

**Comment by CarolEidt** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

File: src/jit/optcse.cpp [hunk 2]
```diff
@@ -1754,6 +1754,22 @@ class CSE_Heuristic
...
+            // This is a non-enregisterable struct.
+            canEnregister                  = false;
+            GenTree*             value     = candidate->Expr();
+            CORINFO_CLASS_HANDLE structHnd = m_pCompiler->gtGetStructHandleIfPresent(candidate->Expr());
+            assert((structHnd != NO_CLASS_HANDLE) || (cseLclVarTyp != TYP_STRUCT));
+            unsigned size = m_pCompiler->info.compCompHnd->getClassSize(structHnd);
+            // Note that this is a bit conservative because it doesn't take into account that we
+            // might use a vector register for struct copies.
+            slotCount = (size + TARGET_POINTER_SIZE - 1) / TARGET_POINTER_SIZE;
+        }
```
Comment (AndyAyersMS) [inline src/jit/optcse.cpp]: "conservative" strikes me as an odd phrasing...

If we overstate costs and create CSEs that aren't as profitable as they might seem, it seems like we're being aggressive, not conservative.

File: src/jit/optcse.cpp [hunk 3]
```diff
@@ -2536,16 +2577,17 @@ bool Compiler::optIsCSEcandidate(GenTree* tree)
...
-    // to use regular assignments, or handling copyObj.
-    if (varTypeIsStruct(type) || type == TYP_VOID)
+    if (type == TYP_VOID)
+    {
+        return false;
+    }
+
+    // If this is a struct type, we can only consider it for CSE-ing if we can get at
+    // its handle, so that we can create a temp.
+    if ((type == TYP_STRUCT) && (gtGetStructHandleIfPresent(tree) == NO_CLASS_HANDLE))
```
Comment (AndyAyersMS) [inline src/jit/optcse.cpp]: I should probably know this by now, but how do we end up with `TYP_STRUCT` and no handle?

File: src/jit/simd.cpp [hunk 4]
```diff
@@ -2967,6 +2967,10 @@ GenTree* Compiler::impSIMDIntrinsic(OPCODE                opcode,
                     op2 = gtCloneExpr(index);
                 }
 
+                // For the non-constant case, we don't want to CSE the SIMD value, as we will just need to store
+                // it to the stack to do the indexing anyway.
+                op1->gtFlags |= GTF_DONT_CSE;
```
Comment (AndyAyersMS) [inline src/jit/simd.cpp]: Is this just a point in time thing (say because we don't value number SIMD ops)?

File: src/jit/morph.cpp [hunk 5]
```diff
@@ -9037,24 +9007,43 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
...
+        // Note that we won't do this straightaway if this is a SIMD type, since it
+        // may be a promoted lclVar (sometimes we promote the individual float fields of
+        // fixed-size SIMD).
+        if (dest->OperGet() == GT_IND)
         {
-            return tree;
+            noway_assert(asgType != TYP_STRUCT);
+            if (varTypeIsStruct(asgType))
+            {
+                destLclVarTree = fgIsIndirOfAddrOfLocal(dest);
```
Comment (mikedn) [inline src/jit/morph.cpp]: Interesting, the `OperIsBlk()` branch above uses `impIsAddressInLocal` rather than `fgIsIndirOfAddrOfLocal`. I suspect that using `impIsAddressInLocal` there is actually incorrect but that's another story.

File: src/jit/optcse.cpp (hunk 2 above)
Comment (CarolEidt) [reply src/jit/optcse.cpp]: I may be confused here, but I think that if we overestimate `slotCount` we make the `yes_cse_cost` higher, so we are less likely to create a CSE. But I've reworded the comment as:
```
            // Note that the slotCount is used to estimate the reference cost, but it may overestimate this
            // because it doesn't take into account that we might use a vector register for struct copies.
```

File: src/jit/optcse.cpp (hunk 3 above)
Comment (CarolEidt) [reply src/jit/optcse.cpp]: I don't recall, but at some point in the not-too-distant past I attempted to fix up all the cases and gave up.

File: src/jit/simd.cpp (hunk 4 above)
Comment (CarolEidt) [reply src/jit/simd.cpp]: No, this is because there's no good way to implement the variable indexing function without storing the operand to memory.

**Comment by CarolEidt** [comment]
> I'm assuming this generates a fair number of diffs, and that you've done some tuning for CSE and similar already.

I did a small amount of tuning; more is probably warranted.

> Can you start characterizing the diffs?

For x64 windows checked (frameworks + tests):

Total bytes of diff: -41406 (-0.04% of base)
4520 total methods with size differences (4236 improved, 284 regressed), 469768 unchanged.

Most of the diffs are due to not marking things as do-not-enregister and additional CSE's. The biggest impacts are in methods that use SIMD or vector types, but there are also some cases where we enregister promoted (non-SIMD) structs that were previously not enregistered.

There are some regressions (of which I looked at many, but not nearly all), due to CSE'd values that cause spill, extra callee-save reg save/restore in prolog/epilog, and a case where we're doing a tail call (local was previously marked address-exposed when it was only address-taken.

File: src/jit/gentree.cpp [hunk 6]
```diff
@@ -16509,15 +16509,45 @@ CORINFO_CLASS_HANDLE Compiler::gtGetStructHandleIfPresent(GenTree* tree)
...
+                                while (fieldSeq->m_next != nullptr)
+                                {
+                                    fieldSeq = fieldSeq->m_next;
+                                }
+                                if (fieldSeq != FieldSeqStore::NotAField() && !fieldSeq->IsPseudoField())
+                                {
+                                    CORINFO_FIELD_HANDLE fieldHnd = fieldSeq->m_fieldHnd;
+                                    CorInfoType fieldCorType = info.compCompHnd->getFieldType(fieldHnd, &structHnd);
+                                    assert(fieldCorType == CORINFO_TYPE_VALUECLASS);
+                                }
```
[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/gentree.cpp]: They're really quite different, and though some logic could be sharable, the ref context seems quite different from the value-type context.

File: src/jit/rationalize.cpp [hunk 7]
```diff
@@ -755,39 +755,18 @@ Compiler::fgWalkResult Rationalizer::RewriteNode(GenTree** useEdge, ArrayStack<G
...
-                        keepBlk = !comp->isAddrOfSIMDType(dataSrc->AsIndir()->Addr());
-                    }
-                }
-            }
-            RewriteSIMDOperand(use, keepBlk);
-        }
-        break;
+            // Rewrite these as GT_IND.
+            assert(node->TypeGet() == TYP_STRUCT || !use.User()->OperIsInitBlkOp());
+            RewriteSIMDOperand(use, false);
```
Comment (mikedn) [inline src/jit/rationalize.cpp]: I think this would be clearer if the call to `RewriteSIMDOperand` is wrapped in `if (varTypeIsSIMD(node))` like in the `GT_IND` case. And the `// Rewrite these as GT_IND.` comment should be inside the `if`.

File: src/jit/gentree.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/gentree.cpp]: Can't we just get rid of this completly? It seems preferrable to keep `GT_OBJ` nodes around and extract the struct handle from those instead of attempting to recover handles from `GT_IND` this way.

[the remark this answers is not in the document]
Comment (CarolEidt) [reply src/jit/gentree.cpp]: I agree that we'd like to get rid of this in the long term, but this code is to deal with cases where we have already made that transformation. I'm going to look into this a bit further, so that I better understand the cases that require this.

[the remark this answers is not in the document]
Comment (mikedn) [reply src/jit/gentree.cpp]: Yes, my comment was targeted at long term. Well, maybe not that long, just only after I'm done with #21705 that makes `GT_OBJ` node "small" and overall cheap.

File: src/jit/rationalize.cpp (hunk 7 above)
Comment (CarolEidt) [reply src/jit/rationalize.cpp]: Done.

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests

File: src/jit/liveness.cpp [hunk 8]
```diff
@@ -2185,11 +2185,10 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
...
                         rhsNode = asgNode->AsBlk()->Data();
                     }
-                    // TODO-1stClassStructs: There should be an else clause here to handle
-                    // the non-block forms of store ops (GT_STORE_LCL_VAR, etc.) for which
-                    // rhsNode is op1. (This isn't really a 1stClassStructs item, but the
-                    // above was added to catch what used to be dead block ops, and that
-                    // made this omission apparent.)
+                    else
+                    {
+                        rhsNode = asgNode->gtGetOp2();
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: Could you please add a comment about what do we expect here?
The old comment said that `rhsNode is op1`, but here you are using `gtGetOp2()`, was the comment wrong?

Also while you are here could you please look at the code that starts this block:
https://github.com/dotnet/coreclr/blob/2520798548b0c414f513aaaf708399f8ef5a4f6c/src/jit/liveness.cpp#L2164-L2172

It looks wrong that we check `if (asgNode == nullptr) ` twice, should it be:
```
if (asgNode == nullptr)
{
  if (addrNode == nullptr)
  {
    asgNode = nextNode;
  }
  if (asgNode == nullptr)
  {
```
or a separate issue?

File: src/jit/morph.cpp [hunk 9]
```diff
@@ -9000,24 +8986,8 @@ GenTree* Compiler::fgMorphOneAsgBlockOp(GenTree* tree)
     bool       isCopyBlock    = asg->OperIsCopyBlkOp();
     bool       isInitBlock    = !isCopyBlock;
 
-    unsigned             size;
+    unsigned             size   = 0;
```
Comment (sandreenko) [inline src/jit/morph.cpp]: Why was it changed? `size` was initialized on all paths before each use (we would have a compiler error if it was not).

File: src/jit/liveness.cpp (hunk 8 above)
Comment (CarolEidt) [reply src/jit/liveness.cpp]: That is a separate issue. I'll open a new issue to fix that, as well as clarifying the code that I added. I don't really want to re-push and add more overload to the already-overloaded CI system.

Comment (CarolEidt) [reply src/jit/liveness.cpp]: Nevermind. The PR testing was stalled anyway, so I pushed with some clarifying comments (and deleting the extra check for asgNode being non-null).

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by mmitche** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

File: src/jit/liveness.cpp (hunk 8 above)
Comment (sandreenko) [reply src/jit/liveness.cpp]: Thank you.

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test
@dotnet-bot test Ubuntu arm Cross Release crossgen_comparison Build and Test

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib - any objections to merging this? I can't seem to re-trigger the "Test Pri0 Linux arm checked" that was "Cancelled" and I'm not optimistic that a re-try will address it.

**Comment by AndyAyersMS** [comment]
No objection...

**Comment by mrange** [comment]
Cool, how can I test this?

**Comment by CarolEidt** [comment]
@mrange - now that this is merged, you can build the coreclr repo and use it. (Instructions here: https://github.com/dotnet/coreclr/blob/master/Documentation/workflow/UsingYourBuild.md) 
Or, you can download the daily (I think) build once this has been incorporated here: https://github.com/dotnet/core-sdk and some advice about finding the commit can be found here: https://github.com/dotnet/core-setup/blob/master/Documentation/project-docs/how-to-track-changes.md#what-git-commit-was-a-particular-binary-built-from
Ping me (or post an issue) if you find problems with the documentation.

File: src/jit/importer.cpp [hunk 10]
```diff
@@ -618,21 +618,8 @@ inline void Compiler::impAppendStmt(GenTree* stmt, unsigned chkLevel)
...
-                        spillGlobEffects = true;
-                    }
-                }
-                else if ((lhs->OperIsBlk() && !lhs->AsBlk()->HasGCPtr()) ||
-                         ((lhs->OperGet() == GT_LCL_VAR) &&
-                          (lvaTable[lhs->AsLclVarCommon()->gtLclNum].lvStructGcCount == 0)))
-                {
-                    // TODO-1stClassStructs: Previously, spillGlobEffects was set to true for
-                    // GT_INITBLK and GT_COPYBLK, but this is overly conservative, and should be
-                    // revisited. (Note that it was NOT set to true for GT_COPYOBJ.)
```
Comment (CarolEidt) [inline src/jit/importer.cpp]: This was where we were previously *always* spilling on a non-COPYOBJ block assignment. This was (clearly) overly conservative, but by eliminating this conservatism, I failed to add the struct case to the code in `impImportBlockCode` (at the `SPILL_APPEND:` label in importer.cpp) (bug #23545, fixed in PR #23570)
